### PR TITLE
Add code for step-based lever.

### PR DESCRIPTION
**Pull request by tess3rac7**

### CODE REVIEW

File: contracts/ReaperStrategyGranary.sol [hunk 1]
```diff
@@ -50,6 +50,7 @@ contract ReaperStrategyGranary is ReaperBaseStrategyv4, IFlashLoanReceiver, UniM
     // Misc constants
     uint16 private constant LENDER_REFERRAL_CODE_NONE = 0;
     uint256 private constant INTEREST_RATE_MODE_VARIABLE = 2;
+    uint256 private constant LEVER_SAFETY_ZONE = 9250;
```
Comment (tess3rac7) [inline contracts/ReaperStrategyGranary.sol:53]: This is much lower than I'd like but if I go 95% or above I keep getting this error: https://github.com/The-Granary/Granary-Protocol-v1/blob/main/contracts/protocol/libraries/logic/ValidationLogic.sol#L181

File: contracts/ReaperStrategyGranary.sol [hunk 2]
```diff
@@ -318,8 +319,46 @@ contract ReaperStrategyGranary is ReaperBaseStrategyv4, IFlashLoanReceiver, UniM
         uint256 desiredBorrow = (realSupply * targetLtv) / (PERCENT_DIVISOR - targetLtv);
 
         if (desiredBorrow > borrow + minLeverageAmount) {
-            _initFlashLoan(desiredBorrow - borrow, INTEREST_RATE_MODE_VARIABLE, DEPOSIT_FL_IN_PROGRESS);
+            uint256 borrowIncrease = desiredBorrow - borrow;
+
+            // flash loan only if gToken has enough want liquidity
+            if (IERC20Upgradeable(want).balanceOf(address(gWant)) >= borrowIncrease) {
```
Comment (tess3rac7) [inline contracts/ReaperStrategyGranary.sol:325]: I tested by commenting out the `if` as well.

File: contracts/ReaperStrategyGranary.sol [hunk 3]
```diff
@@ -318,8 +319,46 @@ contract ReaperStrategyGranary is ReaperBaseStrategyv4, IFlashLoanReceiver, UniM
...
+                }
+            }
+        }
+    }
+
+    /**
+     * @dev Leverages up one step in an attempt to increase borrow by {_totalBorrowIncrease}.
+     *      Returns the actual amount by which borrow was increased.
+     */
+    function _leverUpStep(uint256 _totalBorrowIncrease) internal returns (uint256) {
```
Comment (tess3rac7) [inline contracts/ReaperStrategyGranary.sol:340]: This is a combination of `_leverDownStep()` as well as https://github.com/Byte-Masons/sonne-loop/blob/DAI/contracts/ReaperStrategySonne.sol#L304

File: test/starter-test.js [hunk 4]
```diff
@@ -86,6 +86,7 @@ describe('Vaults', function () {
         {
           forking: {
             jsonRpcUrl: 'https://late-wild-fire.fantom.quiknode.pro/',
+            blockNumber: 53248840,
```
Comment (tess3rac7) [inline test/starter-test.js:89]: Added this old block number because some tests depend on looping to be profitable and it's not currently.

File: contracts/ReaperStrategyGranary.sol (hunk 1 above)
Comment (degenicus) [reply contracts/ReaperStrategyGranary.sol:53]: About how much does this effect the amount of loops needed etc? Is it a big difference?

File: test/starter-test.js [hunk 5]
```diff
@@ -1028,9 +1029,9 @@ describe('Vaults', function () {
       await moveTimeForward(3600);
       await strategy.harvest();
       let vaultBalance = await want.balanceOf(vault.address);
-      expect(vaultBalance).to.equal(ethers.utils.parseEther('0'));
+      expect(vaultBalance).to.equal(ethers.utils.parseEther('1000'));
```
Comment (degenicus) [inline test/starter-test.js:1032]: Why these changes in the expected values? Shouldnt the code produce the same results?

**Review by degenicus — APPROVED** [review]

File: contracts/ReaperStrategyGranary.sol (hunk 2 above)
Comment (tess3rac7) [reply contracts/ReaperStrategyGranary.sol:325]: Yes, when you flash loan with INTEREST_RATE_MODE_VARIABLE and deposit everything back you don't pay any fee and the system opens a debt position for you instead.

File: test/starter-test.js (hunk 5 above)
Comment (tess3rac7) [reply test/starter-test.js:1032]: some tests were borked. strategy was added with 90% allocation yet the tests were expecting 100% of deposited funds to be in the strategy,

File: contracts/ReaperStrategyGranary.sol [hunk 6]
```diff
@@ -318,8 +319,46 @@ contract ReaperStrategyGranary is ReaperBaseStrategyv4, IFlashLoanReceiver, UniM
...
+        (uint256 supply, uint256 borrow) = getSupplyAndBorrow();
+        (, , uint256 threshLtv, , , , , , , ) = IAaveProtocolDataProvider(DATA_PROVIDER).getReserveConfigurationData(
```
[the remark this answers is not in the document]
Comment (tess3rac7) [reply contracts/ReaperStrategyGranary.sol:342]: exactly
